src/apps/translate/default.nix

```diff
@@ -0,0 +1,146 @@
+{
+  # dream2nix deps
+  utils,
+  callNixWithD2N,
+  translateSourceShortcut,
+  coreutils,
+  moreutils,
+  nix,
+  python3,
+  jq,
+  ...
+}:
+utils.writePureShellScriptBin
+"translate"
+[
+  jq
+  coreutils
+  moreutils
+  nix
+  python3
+]
+''
+  cd $WORKDIR
+
+  usage="usage:
+    $0 SOURCE_SHORTCUT TARGET_DIR"
+
+  if [ "$#" -ne 2 ]; then
+    echo "error: wrong number of arguments"
+    echo "$usage"
+    exit 1
+  fi
+
+  source=''${1:?"error: pass a source shortcut"}
+  targetDir=''${2:?"error: please pass a target directory"}
+  targetDir="$(realpath "$targetDir")"
+
+  sourceInfoPath=''${translateSourceInfoPath:-"$TMPDIR/sourceInfo.json"}
+  translateSkipResolved=''${translateSkipResolved:-"0"}
+
+  mkdir -p $targetDir && cd $targetDir
+
+  export dream2nixConfig="{packagesDir=\"./.\"; projectRoot=\"$targetDir\";}"
+
+  # translate the source shortcut
+  ${translateSourceShortcut} $source > $sourceInfoPath
+
+  # collect data for the packages we will resolve
+  resolveDatas="$TMPDIR/resolveData.json"
+  ${callNixWithD2N} eval --json "
+    let
+      data =
+        l.map
+        (
+          p:
+          let
+            resolve = p.passthru.resolve or p.resolve;
+            mkBashEnv = name: value:
+              \"\''${name}=\" + \"'\" + value + \"'\";
+          in
+            if \"$translateSkipResolved\" == \"1\" && resolve.passthru.project ? dreamLock
+            then null
+            else
+              l.concatStringsSep
+              \";\"
+              (
+                (with resolve.passthru.project; [
+                  (mkBashEnv \"name\" name)
+                  (mkBashEnv \"dreamLockPath\" dreamLockPath)
+                  (mkBashEnv \"subsystem\" subsystem)
+                ]) ++ [
+                  (mkBashEnv \"drvPath\" resolve.drvPath)
+                ]
+              )
+        )
+        (
+          l.attrValues
+          (
+            l.removeAttrs
+            (dream2nix.makeOutputs {
+              source = dream2nix.fetchers.fetchSource {
+                source =
+                  l.fromJSON (l.readFile \"$sourceInfoPath\");
+              };
+            }).packages
+            [\"resolveImpure\"]
+          )
+        );
+    in
+      l.unique (l.filter (i: i != null) data)
+  " > $resolveDatas
+
+  # extract source info once here so we can use it later
+  sourceInfo="$(jq '.' -c -r "$sourceInfoPath")"
+
+  # resolve the packages
+  for resolveData in $(jq '.[]' -c -r $resolveDatas); do
+    # extract project data so we can determine where the dream-lock.json will be
+    eval "$resolveData"
+
+    echo "Resolving:: $name (subsystem: $subsystem) (lock path: $dreamLockPath)"
+
+    # build the resolve script and run it
+    nix build --out-link $TMPDIR/resolve $drvPath
+    $TMPDIR/resolve/bin/resolve
+
+    # patch the dream-lock with our source info so the dream-lock works standalone
```

Comment:
Yeah actually, thinking twice, it's probably a better idea to just put it in Nix code anyways, we can always optimize later. So that's what I did. I also added a unit test, so please check that out too.

Reply:
This is wonderful!